### PR TITLE
Revise arm platform notes regarding soft float

### PR DESCRIPTION
**Pull request by thejpster**
This PR updates the Arm microcontroller platform docs to recommend `-fpregs` instead of `+soft-float` as [discussed on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.60-Ctarget-feature.3D.2Bsoft-float.60.20considered.20harmful)

### CODE REVIEW

File: src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md [hunk 1]
```diff
@@ -50,6 +50,9 @@ to use these flags.
 | Cortex-M7F | SP  | Yes | `cortex-m7` | `-fp64`         |
 | Cortex-M7F | DP  | Yes | `cortex-m7` | None            |
 
+**Note:** Never use the `-fpregs` *target-feature* with these `eabihf` targets
+as it will cause compilation units to have different ABIs, which is unsound.
```
Comment (Urgau) [inline src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md:54]:
```suggestion
<div class="warning">

Never use the `-fpregs` *target-feature* with these `eabihf` targets as it
will cause compilation units to have different ABIs, which is unsound.

</div>
```
cf. https://rust-lang.github.io/mdBook/format/mdbook.html#classwarning

File: src/doc/rustc/src/platform-support/thumbv8m.main-none-eabi.md [hunk 2]
```diff
@@ -74,6 +74,9 @@ to use these flags.
 | Cortex-M85  | DP  | Yes | Int       | `cortex-m85`  | `-mve.fp`             |
 | Cortex-M85  | DP  | Yes | Int+Float | `cortex-m85`  | None                  |
 
+**Note:** Never use the `-fpregs` *target-feature* with these `eabihf` targets
+as it will cause compilation units to have different ABIs, which is unsound.
```
Comment (Urgau) [inline src/doc/rustc/src/platform-support/thumbv8m.main-none-eabi.md:78]:
```suggestion
<div class="warning">

Never use the `-fpregs` *target-feature* with these `eabihf` targets as it
will cause compilation units to have different ABIs, which is unsound.

</div>
```

**Comment by ehuss** [comment]
@rustbot ping arm

Can an Arm maintainer review this, please?

Or, cc @RalfJung, since I've noticed the two of you discussing this on Zulip, if you want to review this.

**Comment by thejpster** [comment]
AIUI there are no Arm maintainers for M-Profile, only for the Aarch A-Profile. Although they offered to help on a best-effort basis.

**Comment by jamesmunns** [comment]
CC @adamgrieg @thalesfragoso @newAM

File: src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md [hunk 3]
```diff
@@ -50,6 +50,13 @@ to use these flags.
 | Cortex-M7F | SP  | Yes | `cortex-m7` | `-fp64`         |
 | Cortex-M7F | DP  | Yes | `cortex-m7` | None            |
 
+<div class="warning">
+
+Never use the `-fpregs` *target-feature* with these `eabihf` targets
```
Comment (jacobbramley) [inline src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md:55]: What does "these" refer to? Should it just be "... with `eabihf` targets ...", or perhaps "... with hard-float ABI targets (such as `eabihf`) ..." ?

Comment (thejpster) [reply src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md:55]: It should say 'this target'. Broader words can go in the arm target docs.

**Review by jacobbramley — APPROVED** [review]
This seems to do the right thing (at least for the cases I tried and based on the Zulip discussion), and no alarm bells went off when I discussed it with our LLVM teams.

One caveat we noticed — which is really a separate concern — is that there's no way to specify eabihf with no FP. That sounds nonsensical, but it's potentially useful with integer MVE, which uses the same registers. Clang [achieves that with `+nofp`](https://github.com/llvm/llvm-project/blob/55d51dd9dca8220ffaf9260d56dae9f5c34b7120/clang/lib/Driver/ToolChains/Arch/ARM.cpp#L701), which seems to get [translated into `-mfpu=none`](https://github.com/llvm/llvm-project/blob/55d51dd9dca8220ffaf9260d56dae9f5c34b7120/clang/lib/Driver/ToolChain.cpp#L259), but I can't find a way to control that through Rust.

**Comment by thejpster** [comment]
Does Integer MVE let you use s0 and d0 even without an FPU?! Wild.

**Comment by jacobbramley** [comment]
> Does Integer MVE let you use s0 and d0 even without an FPU?! Wild.

Yes, at least for the few instructions you'd need to use this ABI. I've not done any real work on these devices so can't claim to be an expert, but the [Armv8-M reference manual](https://developer.arm.com/documentation/ddi0553/latest/) says that the necessary `VMOV` instructions are available with either FP or MVE.

The FP/vector register file is overlaid so that `q0` aliases `d1:d0` and `s3:s2:s1:s0`, just like A-class AArch32's Neon registers. You could read `d0` as "the lower 64 bits of `q0`", which might feel more natural in a world without FP hardware. There's a layout diagram under rule R<sub>BBYH</sub>, page 180 in [version DDI0553B.y](https://developer.arm.com/documentation/ddi0553/by/?lang=en) of the manual.

File: src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md (hunk 3 above)
Comment (thejpster) [reply src/doc/rustc/src/platform-support/thumbv7em-none-eabi.md]: Fixed

**Comment by thejpster** [comment]
> > Does Integer MVE let you use s0 and d0 even without an FPU?! Wild.
> 
> Yes, at least for the few instructions you'd need to use this ABI. I've not done any real work on these devices so can't claim to be an expert, but the [Armv8-M reference manual](https://developer.arm.com/documentation/ddi0553/latest/) says that the necessary `VMOV` instructions are available with either FP or MVE.
> 
> The FP/vector register file is overlaid so that `q0` aliases `d1:d0` and `s3:s2:s1:s0`, just like A-class AArch32's Neon registers. You could read `d0` as "the lower 64 bits of `q0`", which might feel more natural in a world without FP hardware. There's a layout diagram under rule RBBYH, page 180 in [version B.y](https://developer.arm.com/documentation/ddi0553/by/?lang=en) of the manual.

OK, I added some notes to mention that this is a potentially valid setup.

**Review by jacobbramley — APPROVED** [review]

**Comment by ehuss** [comment]
Thanks! If anyone has any objections or changes, feel free to post an update.

@bors r+ rollup

**Comment by Kobzol** [comment]
Noise, this only updated documentation.

@rustbot label: +perf-regression-triaged

**Pull request merged**